From 4c4e85d6582e4eb72043f9da7bf4e16ec61ce958 Mon Sep 17 00:00:00 2001
From: Oleksandr Tyshchenko <oleksandr_tyshchenko@epam.com>
Date: Fri, 15 Jul 2022 11:20:53 +0300
Subject: [PATCH 16/62] xen-bus: Set offline if backend's state is
 XenbusStateClosed

Both state (XenbusStateClosed) and online (0) are expected by
toolstack/xl devd to completely destroy the device. But "offline"
is never being set by the backend resulting in timeout during
domain destruction, garbage in Xestore and still running Qemu
instance.

Signed-off-by: Oleksandr Tyshchenko <oleksandr_tyshchenko@epam.com>
---
Fixes for:
libxl: error: libxl_device.c:1357:device_destroy_be_watch_cb: Domain 5:timed out while waiting for
/local/domain/1/backend/qdisk/5/51712 to be removed
---
---
 hw/xen/xen-bus.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/hw/xen/xen-bus.c b/hw/xen/xen-bus.c
index c5085b13c3..8a5eb4dc9d 100644
--- a/hw/xen/xen-bus.c
+++ b/hw/xen/xen-bus.c
@@ -661,6 +661,9 @@ static void xen_device_backend_changed(void *opaque)
         xen_device_backend_set_state(xendev, XenbusStateClosed);
     }
 
+    if (xen_device_backend_get_state(xendev) == XenbusStateClosed)
+        xen_device_backend_set_online(xendev, false);
+
     /*
      * If a backend is still 'online' then we should leave it alone but,
      * if a backend is not 'online', then the device is a candidate
-- 
2.34.1

